<a href="https://colab.research.google.com/github/lyntos/Python_MS_Wilmu_02/blob/main/6_Exercises.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Week 6 Exercises
## Chapter 11
### **Computer Science Thinking:**
### Recursion, Searching, Sorting, and Big O

#### Complete Exercises 11.1 - 11.20


## 11.1
> What does the following code do?
```python
  def mystery(a,b):
    if b == 1:
      return a
    else:
      return a + mystery(a, b - 1)

  mystery(2,10)
```

### 11.1 Answer
> This code passes two integers (a,b) into a recursive function which either returns the integer value of 'a' if b == 1, otherwise it will add a to a, then decrement b by 1 until b == 1. in other words, it creates a recursive, decrementing sum of the same number (a), from the provided value of 'b' through 1 times.

This provided code will add the value of a recursively 10 times, i.e. return 20.

## 11.2
> Find the logic error(s) in the following recursive function, and explain how to correct it (them). This function should find the sum of the values from 0 to n.

```python
  def sum(n):
    if n == 0:
      return 0
    else:
      return n + sum(n)
    
```

### 11.2 Answer
> this will create an infinite recursive call, becuase the value of n is never decremented. It must reflect the code above by changing the **else** code block to:
```python
else:  
  return n + sum(n-1)
```

## 11.3
> What does the following code do?

```python
def mystery(a_array, size):
  if size == 1:
    return a_array[0]
  else:
    return a_array[size-1] + mystery(a_array, size - 1)

import numpy as np

numbers = np.arange(1,11)

mystery(numbers,len(numbers))

```

### 11.3 Answer
> This code sums the values of all the integers in an array, using the same process as 11.1.

## 11.4
> In Section 11.3, we presented a recursive  **factorial** function. What happens if you remove the **if** statement from the **factorial** function, then call the function?

```python
def factorial(number):
  """return factorial of a number"""
  if number <= 1:
    return 1
  return number * factorial(number - 1) # recursive call

```

### 11.4 Answer
> If we remove the **if** statement, the code will reach maximum recursion depth becuase it will never stop - the **if** condition forces the code to return 1 for values <= 1, and by removing that it will continue to $-\infty$

## 11.5 (Recursive **Power** Function)
>  Write a recursive function **power(base, exponent)** that, when called, returns
> $base^{exponent}$

> For example, **power(3,4)** = 3 * 3 * 3 * 3.
> Assume that **exponent** is an integer greater than or equal to 1.

> *Hint*: The recursion step should use the relationship

> $base^{exponent} = base * base ^{exponent - 1}$

> and the terminating condition occurs when **exponent** is equal to 1, because

> $base^1 = base$

> Incorporate this function into a program that enables the user to enter the **base** and **exponent**.

In [ ]:
# part 1 - define the function power()
def power(base,exponent):
  """recursive exponentiation function"""
  if exponent == 1:
    return base
  else:
    return base * power(base,exponent-1)

# part 2 - incorporate into user-input function
user_base = int(input("Enter the base: "))
user_exponent = int(input("Enter the exponent: "))

# show result through power() function
power(user_base,user_exponent)

Enter the base: 2
Enter the exponent: 5


32

## 11.6 (Recursive Fibonacci Modification)

> Modify Section 11.4’s recursive **fibonacci** function to keep track of the total number of function calls. Display the number of calls for **fibonacci(10)**, **fibonacci(20)** and **fibonacci(30)**.

In [ ]:
def fibonacci(n):
  # instantiate calls for this function
  current_calls = 1

  if n in (0, 1):  # base cases
      return n, current_calls
  else:
      # recursively call and get the fib number and calls from sub-problems
      fib_n_minus_1, calls_n_minus_1 = fibonacci(n - 1)
      fib_n_minus_2, calls_n_minus_2 = fibonacci(n - 2)

      # sum the fib numbers and the calls
      total_fib = fib_n_minus_1 + fib_n_minus_2
      total_calls = current_calls + calls_n_minus_1 + calls_n_minus_2

      return total_fib, total_calls

fib_10, calls_10 = fibonacci(10)
print(f"fibonacci(10) = {fib_10:,}\t\t calls = {calls_10:,}")
%timeit -n 1 -r 10 fibonacci(10)

fib_20, calls_20 = fibonacci(20)
print(f"fibonacci(20) = {fib_20:,}\t\t calls = {calls_20:,}")
%timeit -n 1 -r 10 fibonacci(20)

fib_30, calls_30 = fibonacci(30)
print(f"fibonacci(30) = {fib_30:,}\t\t calls = {calls_30:,}")
%timeit -n 1 -r 10 fibonacci(30)

fib_35, calls_35  = fibonacci(35)
print(f"fibonacci(35) = {fib_35:,}\t calls = {calls_35:,}")
%timeit -n 1 -r 10 fibonacci(35)

fibonacci(10) = 55		 calls = 177
22.9 µs ± 707 ns per loop (mean ± std. dev. of 10 runs, 1 loop each)
fibonacci(20) = 6,765		 calls = 21,891
3.21 ms ± 493 µs per loop (mean ± std. dev. of 10 runs, 1 loop each)
fibonacci(30) = 832,040		 calls = 2,692,537
508 ms ± 164 ms per loop (mean ± std. dev. of 10 runs, 1 loop each)
fibonacci(35) = 9,227,465	 calls = 29,860,703
4.41 s ± 677 ms per loop (mean ± std. dev. of 10 runs, 1 loop each)


## 11.7 (Improving Recursive Fibonacci Performance: Memoization)

> Research the performance enhancement technique called memoization. Modify Section 11.4’s recursive  function to incorporate memoization. Compare the performance of both versions for **fibonacci(10)**, **fibonacci(20)** and **fibonacci(30)**.

In [ ]:
# memoized recursive Fibonacci sequence
import numpy as np

def memoized_fibonacci_wrapper(n):
    # initialize memoization table for each top-level call
    fib_memo = np.zeros(n + 1, dtype=int) # Use int for fib numbers

    def _fib_recursive(k):
        if fib_memo[k] != 0: # Check if already computed
            return fib_memo[k]
        if k <= 1:
            fib_memo[k] = k
            return k
        else:
            result = _fib_recursive(k - 1) + _fib_recursive(k - 2)
            fib_memo[k] = result
            return result
    return _fib_recursive(n)

fib_memoized_10 = memoized_fibonacci_wrapper(10)
print(f"fibonacci(10) = {fib_memoized_10:,}")
%timeit -n 1 -r 1000 memoized_fibonacci_wrapper(10)

fib_memoized_20 = memoized_fibonacci_wrapper(20)
print(f"fibonacci(20) = {fib_memoized_20:,}")
%timeit -n 1 -r 1000 memoized_fibonacci_wrapper(20)

fib_memoized_30 = memoized_fibonacci_wrapper(30)
print(f"fibonacci(30) = {fib_memoized_30:,}")
%timeit -n 1 -r 1000 memoized_fibonacci_wrapper(30)

fib_memoized_35 = memoized_fibonacci_wrapper(35)
print(f"fibonacci(35) = {fib_memoized_35:,}")
%timeit -n 1 -r 1000 memoized_fibonacci_wrapper(35)

fibonacci(10) = 55
The slowest run took 167.07 times longer than the fastest. This could mean that an intermediate result is being cached.
15 µs ± 39.2 µs per loop (mean ± std. dev. of 1000 runs, 1 loop each)
fibonacci(20) = 6,765
The slowest run took 19.88 times longer than the fastest. This could mean that an intermediate result is being cached.
21.5 µs ± 11.5 µs per loop (mean ± std. dev. of 1000 runs, 1 loop each)
fibonacci(30) = 832,040
The slowest run took 4.00 times longer than the fastest. This could mean that an intermediate result is being cached.
21.5 µs ± 3.44 µs per loop (mean ± std. dev. of 1000 runs, 1 loop each)
fibonacci(35) = 9,227,465
The slowest run took 4.98 times longer than the fastest. This could mean that an intermediate result is being cached.
27 µs ± 8.26 µs per loop (mean ± std. dev. of 1000 runs, 1 loop each)


## 11.8 (Visualizing Recursion)

> It’s interesting to watch recursion “in action.” Modify the recursive **factorial** function presented in this chapter to print its local variable and recursive-call parameter. For each recursive call, display the outputs on a separate line and add a level of indentation. Make the outputs clear, interesting and meaningful. Your goal here is to design and implement an output format that makes it easier to understand recursion.

In [ ]:
def factorial_visualized(number, indent=0):
  print('  ' * indent + f'Calling factorial_visualized({number})')
  if number <= 1:
    print('  ' * indent + f'Base case: {number} <= 1, returning 1')
    return 1
  else:
    result = number * factorial_visualized(number - 1, indent + 1)
    print('  ' * indent + f'Returning {number} * (factorial_visualized({number - 1})) = {result}')
    return result

print("Visualizing factorial(6):")
factorial_visualized(6)
print("\nVisualizing factorial(3):")
factorial_visualized(3)

Visualizing factorial(6):
Calling factorial_visualized(6)
  Calling factorial_visualized(5)
    Calling factorial_visualized(4)
      Calling factorial_visualized(3)
        Calling factorial_visualized(2)
          Calling factorial_visualized(1)
          Base case: 1 <= 1, returning 1
        Returning 2 * (factorial_visualized(1)) = 2
      Returning 3 * (factorial_visualized(2)) = 6
    Returning 4 * (factorial_visualized(3)) = 24
  Returning 5 * (factorial_visualized(4)) = 120
Returning 6 * (factorial_visualized(5)) = 720

Visualizing factorial(3):
Calling factorial_visualized(3)
  Calling factorial_visualized(2)
    Calling factorial_visualized(1)
    Base case: 1 <= 1, returning 1
  Returning 2 * (factorial_visualized(1)) = 2
Returning 3 * (factorial_visualized(2)) = 6


6

## 11.9 (Greatest Common Divisor)

> The greatest common divisor of integers **x** and **y** is the largest integer that evenly divides into both **x** and **y**. Write and test a recursive function **gcd** that returns the greatest common divisor of **x** and **y** . The **gcd** of **x** and **y** is defined recursively as follows:
      
      If y is equal to 0, then gcd(x,y) is x;
      otherwise,
      gcd(x,y) is gcd(y, x % y), where % is the remainder operator.

In [ ]:
def gcd(x,y):
  if y == 0:
    return x
  else:
    return gcd(y, x % y)

gcd(125,175)

25

## 11.10 (Palindromes)

> A palindrome is a string that is spelled the same way forward and backward. Some examples of palindromes are “radar,” “able was i ere i saw elba” and (if spaces are ignored) “a man a plan a canal panama.”

> Write a recursive **test_palindrome** function that returns **True** if the string stored in an array is a palindrome and  otherwise. The function should ignore spaces and punctuation in the string.

In [ ]:
def test_palindrome(string):
  a_array = list(string.lower())
  b_array = a_array.copy()
  b_array.reverse()
  if a_array == b_array:
    return True


In [ ]:
def test_palindrome_recursive(s):
  '''recursively check for palindrome by comparing s[0] and s[-1]'''
  # convert to lowercase and remove non-alphanumeric characters
  cleaned_s = "".join(char for char in s.lower() if char.isalnum())

  # Base case - empty string or a single-character string is a palindrome
  if len(cleaned_s) <= 1:
    return True
  # recursive step - check if the first and last characters match,
  # then recursively call the function on the substring between them
  elif cleaned_s[0] == cleaned_s[-1]:
    return test_palindrome_recursive(cleaned_s[1:-1])
  # if characters don't match, it's not a palindrome
  else:
    return False

# Test cases
print(f"'radar' is a palindrome: {test_palindrome_recursive('radar')}")
print(f"'able was i ere i saw elba' is a palindrome: {test_palindrome_recursive('able was i ere i saw elba')}")
print(f"'a man a plan a canal panama' is a palindrome: {test_palindrome_recursive('a man a plan a canal panama')}")
print(f"'hello' is a palindrome: {test_palindrome_recursive('hello')}")


'radar' is a palindrome: True
'able was i ere i saw elba' is a palindrome: True
'a man a plan a canal panama' is a palindrome: True
'hello' is a palindrome: False


## 11.11 (Eight Queens)

> A puzzler for chess buffs is the Eight Queens problem, which asks: Is it possible to place eight queens on an empty chessboard so that no queen is “attacking” any other (i.e., no two queens are in the same row, in the same column or along the same diagonal)? For example, if a queen is placed in the upper-left corner of the board, no other queens could be placed in any of the marked squares shown in the following figure. Solve the problem recursively.

> [Hint: Your solution should begin with the first column and look for a location in that column where a queen can be placed—initially, place the queen in the first row. The solution should then recursively search the remaining columns. In the first few columns, there will be several locations where a queen may be placed. Take the first available location. If a column is reached with no possible location for a queen, the program should return to the previous column, and move the queen in that column to a new row. This continuous backing up and trying new alternatives is an example of recursive backtracking.]


In [ ]:
def is_safe(board, row, col):
  # check this row on left side
  for i in range(col):
    if board[row][i] == 1:
      return False

  # check upper diagonal on left side
  for i, j in zip(range(row, -1, -1), range(col, -1, -1)):
    if board[i][j] == 1:
      return False

  # check lower diagonal on left side
  for i, j in zip(range(row, 8, 1), range(col, -1, -1)):
    if board[i][j] == 1:
      return False

  return True

def solve_eight_queens(board, col):
  # base case: if all queens are placed, return true
  if col >= 8:
    return True

  # consider this column and try placing this queen in all rows one by one
  for row in range(8):
    if is_safe(board, row, col):
      # place this queen in board[row][col]
      board[row][col] = 1

      # recur to place the rest of the queens
      if solve_eight_queens(board, col + 1):
        return True

      # if placing queen in board[row][col] doesn't lead to a solution,
      # then remove queen from board[row][col] (backtrack)
      board[row][col] = 0

  # if the queen cannot be placed in any row in this column, return false
  return False

def print_solution(board, show_threats=False):
    for row in board:
      print(" ".join("Q" if x == 1 else "." for x in row))

# initialize the 8x8 chessboard with all zeros
board = [[0 for _ in range(8)] for _ in range(8)]

if solve_eight_queens(board, 0):
  print("solution for eight queens problem:")
  print_solution(board)
else:
  print("solution does not exist")

solution for eight queens problem:
Q . . . . . . .
. . . . . . Q .
. . . . Q . . .
. . . . . . . Q
. Q . . . . . .
. . . Q . . . .
. . . . . Q . .
. . Q . . . . .


## 11.12 (Towers of Hanoi)

> In this chapter, you studied functions that can be easily implemented both recursively and iteratively. In this exercise, we present a problem whose recursive solution demonstrates the elegance of recursion, and whose iterative solution may not be as apparent.
              
>The Towers of Hanoi is one of the most famous classic problems every
    budding computer scientist must grapple with.
    Legend has it that in a temple in the Far East,
    priests are attempting to move a stack of golden disks
    from one diamond peg to another. The following diagram shows
    the pegs with four disks on peg 1.

>The initial stack has 64 disks threaded onto one peg
    and arranged from bottom to top by decreasing size.
    The priests are attempting to move the stack from one peg to another
    under the constraints that exactly one disk is moved at a time
    and at no time may a larger disk be placed above a smaller disk.
    Three pegs are provided, one being used for temporarily holding disks.
    Supposedly, the world will end when the priests complete their task,
    so there is little incentive for us to facilitate their efforts.
              
>Let’s assume that the priests are attempting to move the disks from peg 1 to peg 3.
We wish to develop an algorithm that prints the precise sequence of peg-to-peg disk transfers.
              
>If we were to approach this problem with conventional functions,
    we would rapidly find ourselves hopelessly knotted up in managing
    the disks. Instead, attacking this problem with recursion in mind
    allows the steps to be simple. Moving n disks can be viewed in
    terms of moving only  disks (hence, the recursion), as follows:
              
                
      a. Move n - 1 disks from peg 1 to peg 2,
      using peg 3 as a temporary holding area.
      b. Move the last disk (the largest) from peg 1 to peg 3.           
      c. Move the n - 1 disks from peg 2 to peg 3,
      using peg 1 as a temporary holding area.
              
>The process ends when the last task involves moving n=1 disk (i.e., the base case).
    This task is accomplished by simply moving the disk,
    without the need for a temporary holding area.
    Write a program to solve the Towers of Hanoi problem.
    Use a recursive function with four parameters:
              
                
      a. The number of disks to be moved.       
      b. The peg on which these disks are initially threaded.
      c. The peg to which this stack of disks is to be moved.
      d. The peg to be used as a temporary holding area.
              
>Display the precise instructions for moving the disks from the starting
    peg to the destination peg. To move a stack of three disks from peg 1
    to peg 3, the program displays the following moves:

      1 -> 3
      1 -> 2
      3 -> 2
      1 -> 3
      2 -> 1
      2 -> 3
      1 -> 3

In [ ]:
def towers_of_hanoi(n, source_peg, destination_peg, auxiliary_peg):
  # base case: if only one disk, move it directly
  if n == 1:
    print(f"{source_peg} -> {destination_peg}")
    return

  # move n-1 disks from source to auxiliary, using destination as temp
  towers_of_hanoi(n - 1, source_peg, auxiliary_peg, destination_peg)

  # move the largest disk from source to destination
  print(f"{source_peg} -> {destination_peg}")

  # move n-1 disks from auxiliary to destination, using source as temp
  towers_of_hanoi(n - 1, auxiliary_peg, destination_peg, source_peg)

print("Moves for 3 disks from peg 1 to peg 3:")
towers_of_hanoi(3, 1, 3, 2)

print("\nMoves for 4 disks from peg 1 to peg 3:")
towers_of_hanoi(4, 1, 3, 2)

print("\nMoves for 5 disks from peg 1 to peg 3:")
towers_of_hanoi(5, 1, 3, 2)

Moves for 3 disks from peg 1 to peg 3:
1 -> 3
1 -> 2
3 -> 2
1 -> 3
2 -> 1
2 -> 3
1 -> 3

Moves for 4 disks from peg 1 to peg 3:
1 -> 2
1 -> 3
2 -> 3
1 -> 2
3 -> 1
3 -> 2
1 -> 2
1 -> 3
2 -> 3
2 -> 1
3 -> 1
2 -> 3
1 -> 2
1 -> 3
2 -> 3

Moves for 5 disks from peg 1 to peg 3:
1 -> 3
1 -> 2
3 -> 2
1 -> 3
2 -> 1
2 -> 3
1 -> 3
1 -> 2
3 -> 2
3 -> 1
2 -> 1
3 -> 2
1 -> 3
1 -> 2
3 -> 2
1 -> 3
2 -> 1
2 -> 3
1 -> 3
2 -> 1
3 -> 2
3 -> 1
2 -> 1
2 -> 3
1 -> 3
1 -> 2
3 -> 2
1 -> 3
2 -> 1
2 -> 3
1 -> 3


## 11.13 (Logarithmic Portion of Big O)

> What key aspect of both the binary search and the merge sort accounts for the logarithmic portion of their respective Big Os?

### 11.13 Answer
> The logarithmic portion comes from repeatedly halving the array to be searched.

## 11.14 (Comparing Insertion Sort to Merge Sort)

> In what sense is the insertion sort superior to the merge sort? In what sense is the merge sort superior to the insertion sort?

### 11.14 Answer
> Insertion Sort has a memory complexity of $O(1)$ and a time complexity of $O(n^2)$, becuase it is basically a nested for loop. Insertion Sort is superior for small or mostly sorted data sets

> Merge sort is "far more efficient than insertion or selection sort". Merge sort has a total efficiency of $O(n \log n)$. This is especially useful in most cases outside of those described in the paragraph above, i.e. large data sets

## 11.15 (Merge Sort: Sorting Subarrays)

> In the text, we say that after the merge sort splits the array into two subarrays, it then sorts these two subarrays and merges them. Why might someone be puzzled by our statement that “it then sorts these two subarrays”?

### 11.15 Answer
> This could confuse someone because it is not actually calling a sorting function, rather it is recursively splitting the arrays until it has subarrays of one element. Once it has the arrays of one element, then it merges them back together in the correct order - the sort.

## 11.16 (Timing Sorting Algorithms)

> Remove the output statements from the functions **selection_sort**, **insertion_sort** and **merge_sort** defined in this chapter, then import each example’s source-code file into IPython.

> Create a 100,000-element array of random integers named **data1** and make two additional copies of the array ( **data2** and **data3**) by calling method **copy** on the original array.

> Next, use **%timeit** as follows to compare the performance of each sorting algorithm:
          
    %timeit -n 1 -r 1 selectionsort.selection_sort(data1)
    %timeit -n 1 -r 1 insertionsort.insertion_sort(data2)
    %timeit -n 1 -r 1 mergesort.merge_sort(data3)       
      
> Do the **selection_sort** and **insertion_sort** take approximately the same amount of time?

> Is **merge_sort** much faster?

In [ ]:
# def print_pass(data, pass_number, index):
#     """Print a pass of the algorithm."""
#     label = f'after pass {pass_number}: '
#     print(label, end='')

#     # output elements up to selected item
#     print('  '.join(str(d) for d in data[:index]),
    #     end='  ' if index != 0 else '')

    # print(f'{data[index]}* ', end='')  # indicate swap with *

    # # output rest of elements
    # print('  '.join(str(d) for d in data[index + 1:len(data)]))

    # # underline elements that are sorted after this pass_number
    # print(f'{" " * len(label)}{"--  " * pass_number}')

In [ ]:
"""Sorting an array with selection sort."""
import numpy as np

def selection_sort(data):
    """Sort array using selection sort."""
    # loop over len(data) - 1 elements
    for index1 in range(len(data) - 1):
        smallest = index1  # first index of remaining array

        # loop to find index of smallest element
        for index2 in range(index1 + 1, len(data)):
            if data[index2] < data[smallest]:
                smallest = index2

        # swap smallest element into position
        data[smallest], data[index1] = data[index1], data[smallest]
        # print_pass(data, index1 + 1, smallest)

# def main():
#     data = np.array([35, 73, 90, 65, 23, 86, 43, 81, 34, 58])
#     print(f'Unsorted array: {data}\n')
#     selection_sort(data)
#     print(f'\nSorted array: {data}\n')

# # call main if this file is executed as a script
# if __name__ == '__main__':
#     main()

In [ ]:
import numpy as np

def insertion_sort(data):
    """Sort an array using insertion sort."""
    # loop over len(data) - 1 elements
    for next in range(1, len(data)):
        insert = data[next]  # value to insert
        move_item = next  # location to place element

        # search for place to put current element
        while move_item > 0 and data[move_item - 1] > insert:
            # shift element right one slot
            data[move_item] = data[move_item - 1]
            move_item -= 1

        data[move_item] = insert  # place inserted element
        # print_pass(data, next, move_item)  # output pass of algorithm

# def main():
#     data = np.array([35, 73, 90, 65, 23, 86, 43, 81, 34, 58])
#     print(f'Unsorted array: {data}\n')
#     insertion_sort(data)
#     print(f'\nSorted array: {data}\n')

# # call main if this file is executed as a script
# if __name__ == '__main__':
#     main()

In [ ]:
import numpy as np

# calls recursive sort_array method to begin merge sorting
def merge_sort(data):
    sort_array(data, 0, len(data) - 1)

def sort_array(data, low, high):
    """Split data, sort subarrays and merge them into sorted array."""
    # test base case size of array equals 1
    if (high - low) >= 1:  # if not base case
        middle1 = (low + high) // 2  # calculate middle of array
        middle2 = middle1 + 1  # calculate next element over

        # output split step
        # print(f'split:   {subarray_string(data, low, high)}')
        # print(f'         {subarray_string(data, low, middle1)}')
        # print(f'         {subarray_string(data, middle2, high)}\n')

        # split array in half then sort each half (recursive calls)
        sort_array(data, low, middle1)  # first half of array
        sort_array(data, middle2, high)  # second half of array

        # merge two sorted arrays after split calls return
        merge(data, low, middle1, middle2, high)

# merge two sorted subarrays into one sorted subarray
def merge(data, left, middle1, middle2, right):
    left_index = left  # index into left subarray
    right_index = middle2  # index into right subarray
    combined_index = left  # index into temporary working array
    merged = [0] * len(data)  # working array

    # output two subarrays before merging
    # print(f'merge:   {subarray_string(data, left, middle1)}')
    # print(f'         {subarray_string(data, middle2, right)}')

    # merge arrays until reaching end of either
    while left_index <= middle1 and right_index <= right:
        # place smaller of two current elements into result
        # and move to next space in arrays
        if data[left_index] <= data[right_index]:
            merged[combined_index] = data[left_index]
            combined_index += 1
            left_index += 1
        else:
            merged[combined_index] = data[right_index]
            combined_index += 1
            right_index += 1

    # if left array is empty
    if left_index == middle2:  # if True, copy in rest of right array
        merged[combined_index:right + 1] = data[right_index:right + 1]
    else: # right array is empty, copy in rest of left array
        merged[combined_index:right + 1] = data[left_index:middle1 + 1]

    data[left:right + 1] = merged[left:right + 1]  # copy back to data

    # output merged array
    # print(f'         {subarray_string(data, left, right)}\n')

# method to output certain values in array
def subarray_string(data, low, high):
    temp = '   ' * low  # spaces for alignment
    temp += ' '.join(str(item) for item in data[low:high + 1])
    return temp

# def main():
#     data = np.array([35, 73, 90, 65, 23, 86, 43, 81, 34, 58])
#     print(f'Unsorted array: {data}\n')
#     merge_sort(data)
#     print(f'\nSorted array: {data}\n')

# # call main if this file is executed as a script
# if __name__ == '__main__':
#     main()

In [ ]:
import numpy as np

# create a 100,000 element array of random integers
data1 = np.random.randint(1, 100, size=100000)

data2 = data1.copy()
data3 = data1.copy()



In [ ]:
%timeit -n 1 -r 1 selection_sort(data1)

In [ ]:
%timeit -n 1 -r 1 insertion_sort(data2)

In [ ]:
%timeit -n 1 -r 1 merge_sort(data3)

17 s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


## 11.17 Bucket Sort

> A bucket sort begins with a one-dimensional array of positive integers to be sorted and a two-dimensional array of integers with rows indexed from 0 to 9 and columns indexed from 0 to $n -1$, where n is the number of values to be sorted. Each row of the two-dimensional array is referred to as a bucket. Write a class named **BucketSort** containing a function called **sort** that operates as follows:
              
    a. Place each value of the one-dimensional array into a row of the bucket array, based on the value’s “ones” (rightmost) digit.
    For example, 97 is placed in row 7, 3 is placed in row 3, and 100 is placed in row 0.
    This procedure is called a distribution pass.
                
    b. Loop through the bucket array row by row, and copy the values back to the original array.
    This procedure is called a gathering pass.
    The new order of the preceding values in the one-dimensional array is 100, 3 and 97.
                
    c. Repeat this process for each subsequent digit position (tens, hundreds, thousands, etc.).
    On the second (tens digit) pass, 100 is placed in row 0, 3 is placed in row 0 (because 3 has no tens digit), and 97 is placed in row 9.
    After the gathering pass, the order of the values in the one-dimensional array is 100, 3 and 97.
    On the third (hundreds digit) pass, 100 is placed in row 1, 3 is placed in row 0, and 97 is placed in row 0 (after the 3).
    After this last gathering pass, the original array is in sorted order.
                
> The two-dimensional array of buckets is 10 times the length of the integer array being sorted. This sorting technique provides better performance than a selection and insertion sorts but requires much more memory—the selection and insertion sorts require space for only one additional element of data. This comparison is an example of a space/time trade-off: The bucket sort uses more memory than the selection and insertion sorts, but performs better. This version of the bucket sort requires copying all the data back to the original array on each pass. Another possibility is to create a second two-dimensional bucket array and repeatedly swap the data between the two bucket arrays.

In [ ]:
class BucketSort:
  def __init__(self, data):
    self.data = list(data)

  def get_digit(self, num, digit_position):
    return (num // 10**digit_position) % 10

  def sort(self):
    # determine the maximum number of digits in the array
    max_val = 0
    if self.data:
      max_val = max(self.data)

    max_digits = 0
    if max_val > 0:
      max_digits = len(str(max_val))
    elif self.data and max_val == 0:
      max_digits = 1
    elif not self.data:
        return self.data

    # iterate through each digit position (ones, tens, hundreds, etc.)
    for digit_position in range(max_digits):
      # create 10 buckets
      buckets = [[] for _ in range(10)]

      # distribution pass: place each number into the appropriate bucket
      for number in self.data:
        digit = self.get_digit(number, digit_position)
        buckets[digit].append(number)

      # gathering pass: copy numbers from buckets back to the original array
      self.data.clear() # Clear the original array
      for bucket in buckets:
        for number in bucket:
          self.data.append(number)

    return self.data # return the sorted list

# test an array
arr1 = [35, 73, 90, 65, 23, 86, 43]
bucket_sort = BucketSort(arr1)
sorted_arr1 = bucket_sort.sort()
print(sorted_arr1)
%timeit -n 1 -r 10 bucket_sort.sort()

[23, 35, 43, 65, 73, 86, 90]
6.56 µs ± 2.94 µs per loop (mean ± std. dev. of 10 runs, 1 loop each)


## 11.18 (Recursive Binary Search)

> Modify this chapter’s **binary_search** function to perform a recursive binary search of the array. The function should receive the search key, starting index and ending index as arguments. If the search key is found, return its index in the array. If the search key is not found, return $-1$.

In [ ]:
import numpy as np

def recursive_binary_search(data, key, start, end):
    if start > end:
        return -1

    middle = (start + end) // 2

    # if the key is found at the middle
    if key == data[middle]:
        return middle
    # if the key is in the left half
    elif key < data[middle]:
        return recursive_binary_search(data, key, start, middle - 1)
    # if the key is in the right half
    else:
        return recursive_binary_search(data, key, middle + 1, end)


sorted_data = np.array([2, 5, 8, 12, 16, 23, 38, 56, 72, 91])
print(f"Sorted array: {sorted_data}")

# test cases
key_to_find = 23
index = recursive_binary_search(sorted_data, key_to_find, 0, len(sorted_data) - 1)
print(f"Key {key_to_find} found at index: {index}" if index != -1 else f"Key {key_to_find} not found")


key_to_find = 91
index = recursive_binary_search(sorted_data, key_to_find, 0, len(sorted_data) - 1)
print(f"Key {key_to_find} found at index: {index}" if index != -1 else f"Key {key_to_find} not found")


--- Test Recursive Binary Search ---
Sorted array: [ 2  5  8 12 16 23 38 56 72 91]
Key 23 found at index: 5
Key 91 found at index: 9


## 11.19 (Quicksort)

> The recursive sorting technique called quicksort uses the following basic algorithm for a one-dimensional array of values:
              
                
    a. Partitioning Step: Take the first element of the unsorted array and determine its final location in the sorted array
    (i.e., all values to the left of the element in the array are less than the element,
    and all values to the right of the element in the array are greater than the element
    —we show how to do this below).
    We now have one element in its proper location and two unsorted subarrays.
                
    b. Recursive Step: Perform Step 1 on each unsorted subarray.
    Each time Step 1 is performed on a subarray, another element is placed in its final location of the sorted array, and two unsorted subarrays are created.
    When a subarray consists of one element, that element is in its final location (because a one-element array is already sorted).
    The basic algorithm seems simple enough, but how do we determine the final position of the first element of each subarray?
    As an example, consider the following set of values (the element in bold is the partitioning element—it will be placed in its final location in the sorted array):

>       37 2 6 4 89 8 10 12 68 45           

> The basic algorithm seems simple enough, but how do we determine the final position of the first element of each subarray? As an example, consider the following set of values (the element in bold is the partitioning element—it will be placed in its final location in the sorted array):

> Starting from the rightmost element of the array, compare each element with 37 until an element less than 37 is found; then swap 37 and that element. The first element less than 37 is 12, so 37 and 12 are swapped. The new array is

>       *12* 2 6 4 89 8 10 37 68 45

Element 12 is in "* *" to indicate that it was just swapped with 37.
                
> Starting from the left of the array, but beginning with the element after 12, compare each element with 37 until an element greater than 37 is found—then swap 37 and that element. The first element greater than 37 is 89, so 37 and 89 are swapped. The new array is

>       12 2 6 4 37 8 10 *89* 68 45

> Starting from the right, but beginning with the element before 89, compare each element with 37 until an element less than 37 is found—then swap 37 and that element. The first element less than 37 is 10, so 37 and 10 are swapped. The new array is

>       12 2 6 4 *10* 8 37 89 68 45

> Starting from the left, but beginning with the element after 10, compare each element with 37 until an element greater than 37 is found—then swap 37 and that element. There are no more elements greater than 37, so when we compare 37 with itself, we know that 37 has been placed in its final location in the sorted array. Every value to the left of 37 is smaller than it, and every value to the right of 37 is larger than it.
                
> Once the partition has been applied on the previous array, there are two unsorted subarrays. The subarray with values less than 37 contains 12, 2, 6, 4, 10 and 8. The subarray with values greater than 37 contains 89, 68 and 45. The sort continues recursively, with both subarrays being partitioned in the same manner as the original array.
                
Based on the preceding discussion, write recursive function **quick_sort_helper** to sort a one-dimensional integer array. The function should receive as arguments a starting index and an ending index on the original array being sorted. Call this function from a **quick_sort** function that receives just the original array to sort.


In [ ]:
def quick_sort(data):
  """Initiates the recursive quicksort process."""
  # call the helper function to perform the recursive sort
  if data:
    quick_sort_helper(data, 0, len(data) - 1)

def partition(data, left, right):
  pivot_index = left # choose the first element as the pivot
  pivot_value = data[pivot_index]
  left_pointer = left + 1 # start from the element after the pivot
  right_pointer = right # start from the last element

  done = False
  while not done:
    # move left_pointer to the right until an element > pivot_value is found
    while left_pointer <= right_pointer and data[left_pointer] <= pivot_value:
      left_pointer += 1

    # move right_pointer to the left until an element < pivot_value is found
    while data[right_pointer] >= pivot_value and right_pointer >= left_pointer:
      right_pointer -= 1

    # if right_pointer has crossed left_pointer, then we are done partitioning
    if right_pointer < left_pointer:
      done = True
    else:
      # swap elements at left_pointer and right_pointer
      data[left_pointer], data[right_pointer] = data[right_pointer], data[left_pointer]

  # swap the pivot element with the element at right_pointer's position
  # this places the pivot in its final sorted position
  data[pivot_index], data[right_pointer] = data[right_pointer], data[pivot_index]

  return right_pointer # return the final position of the pivot

def quick_sort_helper(data, first, last):
  if first < last:
    # partition the array and get the pivot's final index
    split_point = partition(data, first, last)

    # recursively sort the two subarrays
    quick_sort_helper(data, first, split_point - 1)
    quick_sort_helper(data, split_point + 1, last)


arr1 = [37, 2, 6, 4, 89, 8, 10, 12, 68, 45]
print(f"Original array: {arr1}")
quick_sort(arr1)
print(f"Sorted array:   {arr1}\n")

## 11.20 (Determining Big O of Various Algorithms)

Determine the Big O for each of the following. You may need to research some of these items online:
              


> a. Get or set an item by index in a Python list.

> b. Insert a new value in order in a Python sorted list.

> c. Shell short an array.

> d. Bubble sort an array.

> e. Towers of Hanoi for n disks. [Hint: For $n = 1,2,3,4,5$ or $6$, the number of operations is 1, 3, 7, 15, 31 or 63, respectively.] You’ll see that the Big O for the Towers of Hanoi—$O(n^2)$—is far worse than that of the $O(2^n)$ sorts in this chapter. By the way, for 64 disks, the number of operations is 18,446,744,073,709,551,615. If the priests could move one disk per second, it would take 584,942,417,355 years to move the stack of 64 disks!

> f. Find all the permutations (unique arrangements) of n unique items. [Hint: For the digits 1, 2 and 3 there are six permutations—123, 132, 213, 231, 312 and 321.] For the digits 1, 2, 3 and 4 there are 24 permutations. For the digits 1, 2, 3, 4 and 5 there are 120 permutations. You’ll see that this Big O is far worse even than that of the Towers of Hanoi.

Item | Time | Space | Note
----|---------|-------|---
a | $O(1)$ | $O(1)$ |
b | $O(n)$ | $O(n)$ | can be $O(1)$
c | $O(n^2)$ | $O(1)$ | Time can be $O(n \log n)$ - depends on gap sequence
d | $O(n^2)$ | $O(1)$ | sorts in place, no extra space needed
e | $O(2^n)$ | $O(n)$ |
f | $O(n!)$ | $O(n! * n)$ |